### ISSUE TITLE
Implement Peer Wire Protocol 

### ISSUE THREAD

**Issue by MatejLach** [MatejLach/rustorrent#5]
This is a tracking issue for implementing the peer wire 

Once [tracker request/response cycle](https://github.com/MatejLach/rustorrent/issues/4) is implemented, we are ready to implement the Peer Wire Protocol.
The Peer Wire Protocol facilitates the actual P2P exchange of blocks and is thus the heart of the BitTorrent protocol. 

- [Specification](https://wiki.theory.org/BitTorrentSpecification#Peer_wire_protocol_.28TCP.29)
- [Another Specification](http://jonas.nitro.dk/bittorrent/bittorrent-rfc.html#anchor20)

**Comment by MatejLach** [comment]
Examples:

- [Example implementation in C#](https://cheatdeath.github.io/research-bittorrent-doc/#peers)
-  [Example python implementation (section 6)](http://www.kristenwidman.com/blog/33/how-to-write-a-bittorrent-client-part-1)

**Comment by synlestidae** [comment]
Examples are definitely worth reading. That C# one - woah! Lots of detail there. 

Don't have a lot of experience with this kind of network programming... but I think [MIO](https://github.com/carllerche/mio) is a good option. One asynchronous event loop would handle reading and writing to peers... not sure what to suggest from there lol but I'll have more ideas soon.

**Comment by MatejLach** [comment]
@synlestidae Yeah, am not too versed in this as well, but that's part of the fun, night :-)
I agree with mio + I wonder whether [futures-rs](https://github.com/alexcrichton/futures-rs) can help us here as well.

**Comment by synlestidae** [comment]
Yes a few months ago I would have said MIO *for sure* as it seemed to be the only (as far as I knew) cross-platform async-IO rust library. There is also [tokio](https://medium.com/@carllerche/announcing-tokio-df6bb4ddb34#.25o4bn51j) which is built on top of MIO. Worth a look...

**Comment by MatejLach** [comment]
I saw that `hyper` is [about to improve its async API by using tokio](https://github.com/tokio-rs/tokio-hyper), however [this benchmark](https://github.com/alexcrichton/futures-rs/tree/master/futures-minihttp), shows that **current** `hyper`, (which does *not* integrate `tokio`), is significantly slower than both `mio` and the `futures` HTTP implementation, so it may be worth seeing how that improves with `hyper` 0.10, implementing `tokio`.
